cheroot/workers/monitor.py

```diff
@@ -0,0 +1,135 @@
+
+import threading
+import time
+
+from dynpool import DynamicPoolResizer
+
+
+class BackgroundTask(threading.Thread):
+    """A subclass of threading.Thread whose run() method repeats.
+
+    Use this class for most repeating tasks. It uses time.sleep() to wait
+    for each interval, which isn't very responsive; that is, even if you call
+    self.cancel(), you'll have to wait until the sleep() call finishes before
+    the thread stops. To compensate, it defaults to being daemonic, which means
+    it won't delay stopping the whole process.
+
+    :param int interval: time interval between invocation of input function
+    :param function: callable that needs to be invoked
+    :param args: list of arguments to pass to the function
+    :param kwargs: dictionary of keyword args to pass to the function
+    """
+
+    def __init__(self, interval, function, name=None, args=None, kwargs=None):
+        """Initialize a BackgroundTask."""
+        super(BackgroundTask, self).__init__(name=name)
+        self.interval = interval
+        self.function = function
+        self.args = args if args is not None else []
+        self.kwargs = kwargs if kwargs is not None else {}
+        self.running = False
+        self.daemon = True
+
+    def cancel(self):
+        """Cancel the task."""
+        self.running = False
+
+    def run(self):
+        """Invoke the callable with the given interval."""
+        self.running = True
+        while self.running:
+            # Any error in the calling function will be raised immediately
+            self.function(*self.args, **self.kwargs)
+            time.sleep(self.interval)
+
+
+class Monitor:
+    """Monitor a BackgroundTask."""
+
+    callback = None
+    """The function to call at intervals."""
+
+    frequency = 60
+    """The time in seconds between callback runs."""
+
+    thread = None
+    """A :class:`BackgroundTask` thread."""
+
+    def __init__(self, callback, frequency=60, name=None):
+        """Initialize Monitor."""
+        self.callback = callback
+        self.frequency = frequency
+        self.thread = None
+        self.name = name or self.__class__.__name__
+
+    def start(self):
+        """Start our callback in its own background thread."""
+        if self.frequency <= 0:
+            raise RuntimeError('The frequency must be a positive number')
+
+        if self.thread is not None:
+            raise RuntimeError('Only one background task can be monitored')
+
+        self.thread = BackgroundTask(
+            self.frequency,
+            self.callback,
+            name=self.name,
+        )
+        self.thread.start()
+
+    def stop(self):
+        """Stop our callback's background task thread."""
+        if self.thread is None:
+            return
+        if self.thread is not threading.currentThread():
+            self.thread.cancel()
+            if not self.thread.daemon:
+                self.thread.join()
+        self.thread = None
+
+    def graceful(self):
```

Comment:
Why is this called graceful and how is it supposed to be used?

Reply:
This is in line with cherryPy states for a background activity. It denotes a graceful restart/reload activity. Not used here as of now, but kept it to be consistent. 
https://github.com/cherrypy/cherrypy/blob/e77f0830754b1eb86ee53d82d10e2c76fe31cae6/cherrypy/process/plugins.py#L577

Reply:
The purpose of having the server in a separate library is so that we don't copy-over unnecessary code from CherryPy keeping it separate. Not sure if this one is needed here. But I'll be happy to hear @jaraco's opinion too...